### ISSUE TITLE
False positive: `too-many-statements-in-try-clause` triggers on try-finally clause

### ISSUE THREAD

**Issue by PerchunPak** [astral-sh/ruff#25390]
### Summary

https://docs.astral.sh/ruff/rules/too-many-statements-in-try-clause/

Sometimes, it is easier to write `try-finally` clauses instead of context managers. Especially if the object is not a context manager, or we need to do other clean-up steps than `session.close()`. Anyway, this rule should not be triggered on a try-finally clause

```py
session = ...
try:
    print()
    print()
    print()
    print()
    print()
    print()
finally:
    session.close()
```

[Playground](https://play.ruff.rs/df009f98-b97c-4d8d-a74f-de06f20a42d8)

### Version

0.15.14

**Comment by ntBre** [comment]
Thanks for the feedback on this preview rule! I'm not sure I follow why the rule shouldn't apply in this case. It sounds more like you may want to use a `noqa` comment in this instance (if you like the rule in general) or simply ignore the rule. Is this a common pattern in your code base?

This also seems consistent with pylint:

```console
$ uvx pylint try.py --load-plugins pylint.extensions.broad_try_clause --disable=all --enable=too-many-try-statements
************* Module try
try.py:2:0: W0717: try clause contains 6 statements, expected at most 1 (too-many-try-statements)
```

**Comment by PerchunPak** [comment]
I am assuming here, but the motivation behind the `try-except` rule is that the clause catches too many statements and is too broad. While `try-finally` has an entirely different purpose. It ensures that the clean-up logic is executed no matter what; it is effectively a syntax sugar for context managers.

I don't really see a reason for limiting the number of statements for clean-up logic, but I see a reason for limiting it for error handling.

> Is this a common pattern in your code base?

In some codebases yes, but in those I usually ignore "function too complex" rules globally. In this particular case, the rule was triggered [here](https://github.com/py-mine/mcstatus/blob/3ea475919306b8e88f646862a264b31a4c1bc278/docs/examples/code/ping_as_java_and_bedrock_in_one_time.py#L33-L48)

**Comment by MichaReiser** [comment]
I understand that the rule only counts the statement in the `try` arm. The statements in the `finally` arm are not counted.

**Comment by PerchunPak** [comment]
That's not what I meant... I meant not the `finally` arm, but the purpose of the entire `try-finally` statement. `try-except`, `try-finally` and `try-except-finally` are different and have different purposes.

**Comment by MichaReiser** [comment]
So you're saying that it should be different rules for a `try-finally` and a `try-except`?

**Comment by PerchunPak** [comment]
Yes. I see this rule useful only for `try-except` and `try-except-finally`

**Comment by MichaReiser** [comment]
I can see this argument. Specifically, the rule's motivation doesn't really apply in that case:

> Try clauses with many statements make unexpected exceptions harder to detect and debug.

Since `try-finally` doesn't expect any diagnostics, it also doesn't matter how many statements the `try` branch contains. 

(The word unexpected in the documentation also feels a bit unfortunate. Isn't this really about a more narrowed scope around expected diagnostics?)

**Comment by ntBre** [comment]
Agreed, it seems reasonable to me to exclude `try-finally`.

I don't really mind the "unexpected" phrasing. I guess I was thinking "narrow the scope to exclude unexpected exceptions" rather than "narrow the scope to include only expected exceptions," but either way works for me. I feel like the second sentence of the `Why is this bad?` section clarifies things anyway:

> Instead, consider narrowing the try clause to only encompass code that may raise exceptions you can anticipate or know about, moving all other statements either before or after the try clause, or factoring out a helper function.

**Comment by ThiefMaster** [comment]
+1 on excluding try..finally

**Comment by beauxq** [comment]
Yes, this should not be reported if there's no `except`

The "why is this bad" reasoning doesn't make sense if it's just `try`/`finally`
